akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -133,6 +135,167 @@ import scala.util.control.NonFatal
   override def toString: String = s"FlattenMerge($breadth)"
 }
 
+
+@InternalApi private[akka] final class PrefixAndDownstream[In, Out, M](n : Int,
+                                                                       f : immutable.Seq[In] => Flow[In, Out, M])
+  extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]]{
+  import shape.{in, out}
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]
+    val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      var accumulated = List.empty[In]
+      var subSource : SubSourceOutlet[In] = null
+      var subSink : SubSinkInlet[Out] = null
+
+      setHandlers(in, out, this)
+
+      override def postStop(): Unit = {
+        matPromise.tryFailure(new AbruptStageTerminationException(this))
+        if(null != subSource && !subSource.isClosed)
+          subSource.complete()
+        if(null != subSink && !subSink.isClosed)
+          subSink.cancel()
+        super.postStop()
+      }
+      /**
+       * Called when the input port has a new element available. The actual element can be retrieved via the
+       * [[GraphStageLogic.grab]] method.
+       */
+      override def onPush(): Unit = {
+        if(null != subSource) {
+          subSource.push(grab(in))
+        } else {
+          accumulated ::= grab(in)
+          if(accumulated.size == n) {
+            materializeFlow()
+          } else {
+            //gi'me some more!
+            pull(in)
+          }
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if(null != subSource) {
+          subSource.complete()
```

Comment:
I think a better style than checking if there is a value all over in the callbacks would be to have an initial In/OutHandler that is not initialized and then switch over to a separate handler when the sub stream has been materialized (we know it will never switch back to not being initialized, so no further checks needed then)

Reply:
we'd still need to check in postStop